man/startctf.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "startctf" "1" "November 26, 2022" "startctf 2.0.0" "User Commands"
.hy
.SH NAME
.PP
startctf - Create a CTF template and automate some common tasks
.SH SYNOPSIS
.PP
\f[B]startctf\f[R] [OPTION(s)]\&...
[NAME] [--version] [--help, -h] [--install-manpage]
.SH DESCRIPTION
.PP
\f[B]startctf\f[R] is a small utility that creates template
directory/file structures to help you organize your CTF challenges.
It also automates some common tasks such as running network scans (using
tools like nmap, rustscan, etc.)
and fuzzing web applications (using tools like ffuf, gobuster, etc.).
.SH OPTIONS
.TP
\f[B]--install-manpage\f[R]
Download and installs the manpage (from the GitHub repository).
Requires root privileges.
.TP
\f[B]-f, --force\f[R]
Forcefully overwrite existing files/directories.
(Caution: This option is dangerous and can cause data loss.)
.TP
\f[B]--ip <IP>\f[R]
Specify the IP address of the target machine.
This option isn\[cq]t required, but scanning/fuzzing/running tools
won\[cq]t work.
.TP
\f[B]-nS, --nmap-scan\f[R]
Run a nmap scan on the target machine.
Requires the \f[B]--ip\f[R] option.
.TP
\f[B]-as, --auto-scan\f[R]
Try to detect which services are running on the target machine and run
the appropriate scan.
Requires the
\f[B]--ip\f[R] and either \f[B]-nS\f[R] or \f[B]-nSA\f[R] options.
.TP
\f[B]-s, --silent\f[R]
Don\[cq]t show any output from any of the tools or the script itself.
.TP
\f[B]-st, --silent-tools\f[R]
Only silence output from tools, but still show status messages.
Will also prevent xterm windows from opening
containing tool output.
.TP
\f[B]-nw --no-warnings\f[R]
Don\[cq]t show any warnings.
.TP
\f[B]-h, --help\f[R]
Show the help message and exit
.TP
\f[B]-V, --version\f[R]
Show version information and exit
.SS NMAP OPTIONS
.PP
Options specific to nmap scans.
These options are only used if the \f[B]-nS\f[R] or \f[B]-nSA\f[R]
options are used.
\f[B]--ip\f[R] is required.
.TP
\f[B]-nSA, --nmap-all\f[R]
Run a full nmap scan on the given IP (all ports, slow.)
.TP
\f[B]--nmap-Pn\f[R]
Don\[cq]t ping the target (pass the \f[B]-Pn\f[R] option to nmap)
.TP
\f[B]-nSV, --no-sV\f[R]
Don\[cq]t detect the version of services running on the target machine
(don\[cq]t pass the \f[B]-sV\f[R] option to nmap.)
.TP
\f[B]-nmap-ep, --nmap-exclude-ports <PORTS>\f[R]
Ports to exclude from the nmap scan.
\f[B]<PORTS>\f[R] should be a comma-separated list of ports or port
ranges (or a
single number.)
.TP
\f[B]--nmap-args <NMAP_ARGS>\f[R]
Pass additional arguments to nmap.
\f[B]<NMAP_ARGS>\f[R] should be a string of arguments.
.SS FFUF OPTIONS
.PP
Options specific to ffuf scans.
These options are only used if the \f[B]-as\f[R] option is used.
.TP
\f[B]--ffuf-args <FFUF_ARGS>\f[R]
Pass additional arguments to ffuf.
\f[B]<FFUF_ARGS>\f[R] should be a string of arguments.
.SS GOBUSTER OPTIONS
.PP
Options specific to gobuster scans.
These options are only used if the \f[B]-as\f[R] option is used.
.TP
\f[B]--gobuster-args <GOBUSTER>\f[R]
Pass additional arguments to gobuster.
\f[B]<GOBUSTER_ARGS>\f[R] should be a string of arguments.
.SH EXAMPLES
.PP
\f[B]TODO: Add examples\f[R]
.SH AUTHOR
.PP
\f[B]startctf\f[R] was written by \f[B]stautonico\f[R].
You can find the source code on
GitHub (https://github.com/stautonico/startctf)
.SH BUGS
.PP
If you find any bugs, please report them on GitHub
issues (https://github.com/stautonico/startctf/issues).
.SH LICENSE
.PP
\f[B]startctf\f[R] is licensed under the MIT license.
See the LICENSE file for more information.
(https://github.com/stautonico/startctf/blob/main/LICENSE)


startctf.py
#!/usr/bin/env python3
import argparse
import os
from configparser import ConfigParser
from datetime import datetime
from threading import Thread
from xml.dom import minidom

current_user = os.getlogin()
home_dir = os.popen(f"eval ech3"
                    f"o ~{current_user}").read().replace("\n", "")

config_dir = os.path.join(home_dir, ".startctf")

config = ConfigParser()

config.read(os.path.join(config_dir, "ctf.conf"))

arg_parser = argparse.ArgumentParser(description="Create a CTF template")

arg_parser.add_argument("name", type=str, help="The name of the ctf (also the name of the created dir)")

arg_parser.add_argument("--ip", help="The ip address of the target server")

arg_parser.add_argument("-nS", "--net-scan", help="Run a simple nmap scan", action="store_true")
arg_parser.add_argument("-nSA", "--net-scan-all", help="Run an nmap scan and scan all ports", action="store_true")
arg_parser.add_argument("-Pn", help="Includes the -Pn flag to the nmap scan (skip ping)", action="store_true")

arg_parser.add_argument("-e4l", "--enum4linux", help="Run an enum4linux scan on the host", action="store_true")


def create_template(name):
    # First check if the directory exists
    if os.path.exists(f"./{name}"):
        print(f"Folder with the name {name} already exists")
        exit(1)

    # Make the directory that will hold the ctf files
    os.mkdir(f"./{name}")

    os.chdir(f"./{name}")

    # Find the user's network mapper (nmap by default)
    network_scanner = config.get("tools", "net_scanner").split("/")[-1]

    # Create the directory that holds the net map scans
    os.mkdir(network_scanner)

    # Make the xml directory (for xml out/internal parsing)
    os.mkdir(os.path.join(network_scanner, "xml"))

    # Make the directory for storing logs (from enum tools etc)
    os.mkdir("logs")

    # Make the directory for exfiltrated documents
    os.mkdir("exfiltrated_docs")

    # Make directory for scripts/exploits/etc
    os.mkdir("scripts")


def create_readme(name, ip=None):
    person_name = config.get("title", "name")
    date = datetime.now().strftime("%b/%d/%Y")
    readme = f"""#### {person_name} - {date}\n
# {name}
---\n

### Files
---\n

### Creds
---\n
"""

    if ip:
        readme += f"\n### Info\n---\nIP Address: {ip}\n"

    with open("README.md", "w") as f:
        f.write(readme)


def run_nmap(args):
    os.popen(" ".join(args)).read()


def nmap_scan(ip, mode="default", pn=False):
    nmap_args = ["nmap", "-sV", "-v", ip]
    if pn:
        nmap_args.append("-Pn")
    if mode == "default":
        nmap_args.append("-oN")
        nmap_args.append("nmap/initial")
        # Also output the xml for internal processing
        nmap_args.append("-oX")
        nmap_args.append("nmap/xml/initial.xml")
        thread = Thread(target=run_nmap, args=[nmap_args])
        thread.start()
        return thread
    if mode == "all":
        nmap_args.append("-A")
        nmap_args.append("-p-")
        nmap_args.append("-oN")
        nmap_args.append("nmap/all_ports")
        # Also output the xml for internal processing
        nmap_args.append("-oX")
        nmap_args.append("nmap/xml/all_ports.xml")
        thread = Thread(target=run_nmap, args=[nmap_args])
        thread.start()
        return thread


def parse_nmap_output(output_file):
    result = minidom.parse(output_file)
    ports = result.getElementsByTagName("port")

    with open("README.md", "a") as f:
        f.write("\n### Ports\n---\n")
        # Write the headers
        f.write("| Port | Status | Service |\n")
        f.write("|:----:|:------:|:-------:|\n")
        for port in ports:
            state = port.childNodes[0]
            service = port.childNodes[1]

            port_number = port.attributes["portid"].value
            port_status = state.attributes["state"].value
            service_name = service.attributes["name"].value

            f.write(f"| {port_number} | {port_status} | {service_name} |\n")


def enum_4_linux(ip):
    os.popen(f"enum4linux -a {ip} | tee logs/enum4linux.log").read()


if __name__ == '__main__':
    has_ip = False
    running_nmap = False
    default_nmap_thread = None
    all_nmap_thread = None
    enum_4_linux_thread = None

    args = arg_parser.parse_args()

    create_template(args.name)
    try:
        create_readme(args.name, args.ip)
        has_ip = True
    except AttributeError:
        create_readme(args.name)

    # Check if we should run nmap scans
    if has_ip:
        if args.net_scan:
            default_nmap_thread = nmap_scan(args.ip, pn=args.Pn)
            running_nmap = True

        if args.net_scan_all:
            all_nmap_thread = nmap_scan(args.ip, mode="all", pn=args.Pn)
            running_nmap = True
    else:
        if args.net_scan or args.net_scan_all:
            print("Unable to run nmap scans, no IP provided")

    if args.enum4linux:
        enum_4_linux_thread = Thread(target=enum_4_linux, args=[args.ip])
        enum_4_linux_thread.start()

    if running_nmap:
        print("Waiting for nmap scan(s) to finish")
        if default_nmap_thread:
            default_nmap_thread.join()
            # Parse the output once its one
            if not all_nmap_thread:
                parse_nmap_output("nmap/xml/initial.xml")
        if all_nmap_thread:
            all_nmap_thread.join()
            # Parse the output once its one
            parse_nmap_output("nmap/xml/all_ports.xml")

    if enum_4_linux_thread:
        print(f"Waiting for enum4linux to finish")
        enum_4_linux_thread.join()


# TODO: add new section to the README.md (Web directories)
